src/wallet/rpcwallet.cpp

```diff
@@ -3912,16 +3912,19 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
         }
     }
 
-    CBlockIndex *stopBlock = pwallet->ScanForWalletTransactions(pindexStart, pindexStop, reserver, true);
-    if (!stopBlock) {
-        if (pwallet->IsAbortingRescan()) {
-            throw JSONRPCError(RPC_MISC_ERROR, "Rescan aborted.");
-        }
-        // if we got a nullptr returned, ScanForWalletTransactions did rescan up to the requested stopindex
+    const CBlockIndex* stopBlock = nullptr;
+    CWallet::ScanResult result =
+        pwallet->ScanForWalletTransactions(pindexStart, pindexStop, reserver, stopBlock, true);
+    switch (result) {
+    case CWallet::ScanResult::SUCCESS:
         stopBlock = pindexStop ? pindexStop : pChainTip;
-    }
-    else {
+        break;
+    case CWallet::ScanResult::FAILURE:
         throw JSONRPCError(RPC_MISC_ERROR, "Rescan failed. Potentially corrupted data files.");
+    case CWallet::ScanResult::USER_ABORT:
+        throw JSONRPCError(RPC_MISC_ERROR, "Rescan aborted.");
+    default:
+        throw JSONRPCError(RPC_MISC_ERROR, strprintf("Unexpected rescan result: %i.", static_cast<int>(result)));
```

Comment:
Should remove the default case, so instead of a runtime-error the compiler can warn about missing cases.
```suggestion
        // no default case, so the compiler can warn about missing cases
```